src/tqec/sketchup/block_graph.py

```diff
@@ -562,3 +584,45 @@ def display(
             filepath_or_bytes=bytes_buffer.getvalue(),
             write_html_filepath=write_html_filepath,
         )
+
+    def get_abstract_observables(self) -> list[AbstractObservable]:
+        """Get all the abstract observables from the block graph."""
+        self.check_validity(allow_virtual_node=False)
+        correlation_subgraphs = self.to_zx_graph().find_correlation_subgraphs()
+        abstract_observables: list[AbstractObservable] = []
+
+        def is_measured(cube: Cube) -> bool:
+            return self.get_pipe(cube.position, cube.position.shift_by(0, 0, 1)) is None
+
+        for g in correlation_subgraphs:
+            if g.num_nodes == 1:
+                cube = ty.cast(Cube, self.get_cube(g.nodes[0].position))
+                abstract_observables.append(
+                    AbstractObservable(frozenset({cube}), frozenset())
+                )
+                continue
+            top_lines: set[Cube | Pipe] = set()
+            bottom_regions: set[Pipe] = set()
+            for edge in g.edges:
+                pipe = ty.cast(Pipe, self.get_pipe(edge.u.position, edge.v.position))
+                u, v = pipe.u, pipe.v
+                if pipe.direction == Direction3D.Z:
+                    if is_measured(v):
+                        top_lines.add(v)
+                    continue
```

Comment:
Is there a reason why this can only be `v` and not `u`? The underlying type is `nx.Graph` and not `nx.DiGraph`, so at first sight I expect edges to also be undirected (i.e., `u` and `v` are in an arbitrary direction).

Reply:
The pipe's `u`/`v` is always sorted by the position so that `u.position < v.position` in `Pipe.__post_init__` method.